builder/dockerfile/shell_parser.go

```diff
@@ -73,36 +84,23 @@ func (sw *shellWord) processStopOn(stopChar rune) (string, error) {
 	return result, nil
 }
 
-func (sw *shellWord) peek() rune {
-	if sw.pos == len(sw.word) {
-		return '\000'
-	}
-	return rune(sw.word[sw.pos])
-}
-
-func (sw *shellWord) next() rune {
-	if sw.pos == len(sw.word) {
-		return '\000'
-	}
-	ch := rune(sw.word[sw.pos])
-	sw.pos++
-	return ch
-}
-
 func (sw *shellWord) processSingleQuote() (string, error) {
 	// All chars between single quotes are taken as-is
 	// Note, you can't escape '
 	var result string
 
-	sw.next()
+	if sw.scanner.Next() == scanner.EOF {
+		return result, nil
+	}
 
 	for {
-		ch := sw.next()
-		if ch == '\000' || ch == '\'' {
+		ch := sw.scanner.Next()
+		if ch == '\000' || ch == '\'' || ch == scanner.EOF {
```

Comment:
Will Next() ever return \000 ?  If not we can remove that from the "if".  If so, I can't help but think we should treat it like any other char and keep going.